lib/Backend/BackwardPass.cpp

```diff
@@ -1818,6 +1908,15 @@ BackwardPass::ProcessBailOutInfo(IR::Instr * instr)
             BVSparse<JitArenaAllocator> * byteCodeUpwardExposedUsed = byteCodeUsesInstr->byteCodeUpwardExposedUsed;
             if (byteCodeUpwardExposedUsed != nullptr)
             {
+                //Stack Args for Formals optimization.
```

Comment:
Actually, I remember why this code is required. There will be some paths which restores scope obj and some path which doesn't restore. So while creating heap arguments object -> we need do not know if the scope object is being restored or not in RestoreValues().  I had two options to tackle this -> either restore scope object in all paths or just recreate it always during the heap arguments creation  (i implemented the second way). 
So it is not guaranteed that there won't be other uses, but it is guaranteed that it will be restored correctly. 
Also, until we complete the forward pass, we do not if we are going to do the optimization or not. That is the reason, why i had to this in the dead store pass.
That's why I always clear the bit here.

I didn't understand your other proposal though.

Let me know if this is unclear.

---
In reply to: [68478772](https://github.com/Microsoft/ChakraCore/pull/1135#discussion_r68478772) [](ancestors = 68478772,68476633,68475602,67800813,67740116)